Add partitioning by field to BigQuery

jeremywadsack opened the pull request:
In #1646 we added support for setting up partitioning when creating tables in BigQuery. About a month later Google released an update to support partitioning by field, rather than just partitioning on the ingestion date (or writing directly to a partition).

This updates `google-cloud-bigquery` gem to support setting that field.

quartzmo commented:
@jeremywadsack Thanks for this PR!

quartzmo commented on google-cloud-bigquery/acceptance/bigquery/table_reference_test.rb:
```diff
@@ -117,6 +118,30 @@
     partitioned_table.reload!
     partitioned_table.table_id.must_equal partitioned_table_id
     partitioned_table.time_partitioning_type.must_equal "DAY"
+    partitioned_table.time_partitioning_fied.must_be_nil
```
NoMethodError: undefined method `time_partitioning_fied' for #<Google::Cloud::Bigquery::Table:0x00007fdcbb5b4e98>
Did you mean?  time_partitioning_field

quartzmo commented on google-cloud-bigquery/acceptance/bigquery/table_reference_test.rb:
```diff
@@ -117,6 +118,30 @@
+    partitioned_table.reload!
+    partitioned_table.table_id.must_equal partitioned_table_id
+    partitioned_table.time_partitioning_type.must_equal "DAY"
+    partitioned_table.time_partitioning_fied.must_equal "dob"
```
NoMethodError: undefined method `time_partitioning_fied' for #<Google::Cloud::Bigquery::Table:0x00007fdcbb5b4e98>
Did you mean?  time_partitioning_field

quartzmo commented on google-cloud-bigquery/acceptance/bigquery/table_test.rb:
```diff
@@ -43,7 +43,11 @@
         updater.time_partitioning_expiration = seven_days
+        updater.schema do |schema|
+          schema.timestamp "dob",   description: "dob description",   mode: :required
+        end
```
There's something wrong here, I get:

```sh
  1) Error:
Google::Cloud::Bigquery::Table::bigquery#test_0007_allows tables to be created with time_partioning enabled:
Google::Cloud::InvalidArgumentError: invalid: Invalid value for: dob is not a valid value
    /Users/quartzmo/code/google/codez/google-cloud-ruby/google-cloud-bigquery/lib/google/cloud/bigquery/service.rb:690:in `rescue in execute'
    /Users/quartzmo/code/google/codez/google-cloud-ruby/google-cloud-bigquery/lib/google/cloud/bigquery/service.rb:684:in `execute'
    /Users/quartzmo/code/google/codez/google-cloud-ruby/google-cloud-bigquery/lib/google/cloud/bigquery/service.rb:161:in `insert_table'
    /Users/quartzmo/code/google/codez/google-cloud-ruby/google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb:507:in `create_table'
    /Users/quartzmo/code/google/codez/google-cloud-ruby/google-cloud-bigquery/acceptance/bigquery/table_test.rb:44:in `block (2 levels) in <top (required)>'
```

quartzmo replied:
What was the issue here and how did you fix it? After the squash I can't see the update.

jeremywadsack commented:
@quartzmo So sorry about the acceptance test failures/typos. I forgot I needed to run those separately. That was sloppy of me.

I've fixed up the test issues and squashed.

quartzmo commented:
@jeremywadsack No problem! You can run everything all at once with `rake ci:a` (as you may know.) Personally I like to develop these features by writing and running the acceptance tests first, just to have the early satisfaction of knowing that the new API feature works as advertised.

I pulled the latest, ran `rake ci:a` and everything passes.

jeremywadsack commented:
Oh, sorry about the preemptive squash. The issue was that I had
time_partitioning_type set to "dob" rather than time_partitioning_field.

Running the acceptance test first makes sense. I'm still wrapping my head
around the gem structure, so I tend to make the change to the code, pull
the local gem into the code I'm working on and integration-test that.
Acceptance tests are pretty slow, but I figured out how to run a single
file instead of the suite.
On Wed, Feb 21, 2018 at 8:23 AM Chris Smith <notifications@github.com>
wrote:

> *@quartzmo* commented on this pull request.
> ------------------------------
>
> In google-cloud-bigquery/acceptance/bigquery/table_test.rb
> <https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/1978#discussion_r169693378>
> :
>
> >          updater.time_partitioning_expiration = seven_days
> +        updater.schema do |schema|
> +          schema.timestamp "dob",   description: "dob description",   mode: :required
> +        end
>
> What was the issue here and how did you fix it? After the squash I can't
> see the update.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/1978#discussion_r169693378>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOurJ6885PKNT2uFhP0XuYXMCsukcv6ks5tXEK8gaJpZM4SLRwQ>
> .
>
-- 
Jeremy Wadsack

quartzmo commented:
For running a single (or multiple) acceptance test(s), `minitest-focus` is already loaded in the project, just put `focus` on the line before the test(s) you want to run, as [shown here](https://github.com/seattlerb/minitest-focus/blob/master/README.txt#L28).
